### PR TITLE
Migration codestyle from Flake8 to ruff

### PR DESCRIPTION
**Pull request by ChrisLovering**
To see these code changes commit-by-commit see [this diff](https://github.com/python-discord/bot/compare/main...ruff-investigation).

This migrates our linting toolchain from flake8, and related plug-ins, to [ruff](https://github.com/charliermarsh/ruff). from testing, ruff is considerably faster, as well as having more auto-fix support for many of the rules.

There are also more rules than what we have enabled in this PR, which we may look into adding in the future.

See the `select` like in pyproject.toml to see the list of rules that have been added. Specific rules have also been disabled, for not fitting into our code style, see `ignore` for which rules have been ignored. A full list of ruff rules can be [found here](https://beta.ruff.rs/docs/rules/)

### CODE REVIEW

File: bot/exts/filtering/_filter_lists/antispam.py
Comment (mbaruh) [inline bot/exts/filtering/_filter_lists/antispam.py:101]: Not necessary

File: bot/exts/filtering/_settings_types/validations/channel_scope.py
Comment (mbaruh) [inline bot/exts/filtering/_settings_types/validations/channel_scope.py:37]: 🥲

File: bot/exts/filtering/_settings_types/validations/bypass_roles.py
Comment (mbaruh) [inline bot/exts/filtering/_settings_types/validations/bypass_roles.py:15]: Same as above

File: bot/exts/filtering/_ui/filter.py
Comment (mbaruh) [inline bot/exts/filtering/_ui/filter.py:391]: This is undesirable here since the variable name helps keep track of the code

File: bot/exts/filtering/_ui/filter_list.py
Comment (mbaruh) [inline bot/exts/filtering/_ui/filter_list.py:29]: Same here

File: bot/exts/filtering/_ui/search.py
Comment (mbaruh) [inline bot/exts/filtering/_ui/search.py:37]: Same here

File: bot/exts/filtering/_utils.py
Comment (mbaruh) [inline bot/exts/filtering/_utils.py:177]: This deletes comments

File: bot/exts/filtering/filtering.py
Comment (mbaruh) [inline bot/exts/filtering/filtering.py:1030]: This isn't going to work, the class expects a function.

File: bot/exts/filtering/_filter_lists/antispam.py
Comment (ChrisLovering) [reply bot/exts/filtering/_filter_lists/antispam.py:101]: Do you think it would be worth a noqa just for this? The rule is to standardise symbols to ascii, so replacing the en dash with a hyphen

File: bot/exts/filtering/_settings_types/validations/channel_scope.py
Comment (ChrisLovering) [reply bot/exts/filtering/_settings_types/validations/channel_scope.py:37]: This is part of the [pyupgrade](https://beta.ruff.rs/docs/rules/#pyupgrade-up) rule, specifically `UP007`. Personally I think having one consistent way here is better, so choosing the less verbose one, that doesn't require an import is the way I'd go too.

File: bot/exts/filtering/filtering.py
Comment (ChrisLovering) [reply bot/exts/filtering/filtering.py:1030]: Updated in https://github.com/python-discord/bot/pull/2523/commits/fae7cea3a948d170ca0fd5e5dfa6cf4e69054b5e

File: bot/exts/filtering/_utils.py
Comment (ChrisLovering) [reply bot/exts/filtering/_utils.py]: Updated in https://github.com/python-discord/bot/pull/2523/commits/6347cf6c1337ab97224263c6891531ce97c0e169

File: bot/exts/filtering/_ui/filter.py
Comment (ChrisLovering) [reply bot/exts/filtering/_ui/filter.py]: https://github.com/python-discord/bot/pull/2523/commits/33f4a76ca74d58201a582d4dd838d142eed4d6fb

File: bot/exts/filtering/_ui/search.py
Comment (ChrisLovering) [reply bot/exts/filtering/_ui/search.py]: https://github.com/python-discord/bot/pull/2523/commits/33f4a76ca74d58201a582d4dd838d142eed4d6fb

File: bot/exts/filtering/_ui/filter_list.py
Comment (ChrisLovering) [reply bot/exts/filtering/_ui/filter_list.py]: https://github.com/python-discord/bot/pull/2523/commits/33f4a76ca74d58201a582d4dd838d142eed4d6fb

**Comment by ChrisLovering** [comment]
Will squash all fixup commits before merging, so marking as do not merge so I can do that once approved.

File: bot/exts/filtering/_settings_types/validations/channel_scope.py
Comment (mbaruh) [reply bot/exts/filtering/_settings_types/validations/channel_scope.py:37]: Sure, but did you read the comment above :D

File: bot/exts/filtering/_filter_lists/antispam.py
Comment (mbaruh) [reply bot/exts/filtering/_filter_lists/antispam.py:101]: We can do a noqa, but so far all the rule has done is make us as noqas

File: bot/exts/filtering/_settings_types/validations/bypass_roles.py
Comment (ChrisLovering) [reply bot/exts/filtering/_settings_types/validations/bypass_roles.py]: Reverted in https://github.com/python-discord/bot/pull/2523/commits/65ec4bfe0404d6081e377c274cd34274e7eeb0db

File: bot/exts/filtering/_settings_types/validations/channel_scope.py
Comment (ChrisLovering) [reply bot/exts/filtering/_settings_types/validations/channel_scope.py]: Reverted in https://github.com/python-discord/bot/pull/2523/commits/65ec4bfe0404d6081e377c274cd34274e7eeb0db

File: bot/exts/filtering/_filter_lists/antispam.py
Comment (ChrisLovering) [reply bot/exts/filtering/_filter_lists/antispam.py:101]: I'd prefer to keep it that way. Non-standard character sets aren't always obvious when reviewing PRs, so they could unknowingly get into the code base and cause confusion/issues. I'd rather have to noqa as needed, and in instances such as this just follow the rule.

File: bot/rules/mentions.py
Comment (mbaruh) [inline bot/rules/mentions.py:1]: Uuuuh this entire file should be gone... must have survived a merge conflict.

**Review by mbaruh — CHANGES_REQUESTED** [review]
There are various places across the codebase where a type alias is set to some Union, which doesn't seem to be handled by the autofixes. You can find them by regex matching ` = (t\.|typing\.)?Union` across the project.

File: bot/rules/mentions.py
Comment (mbaruh) [reply bot/rules/mentions.py:1]: Ignore this, handled in #2528

File: .pre-commit-config.yaml
Comment (ChrisLovering) [reply .pre-commit-config.yaml:18]: Yea, we should do that

Comment (ChrisLovering) [reply .pre-commit-config.yaml]: https://github.com/python-discord/bot/pull/2523/commits/e74e2e50469be60664afd564eab3a6d50961c0f4

File: bot/exts/info/python_news.py
Comment (ChrisLovering) [reply bot/exts/info/python_news.py:112]: Yea, I had a comment in my investigation branch, but didn't cherry pick it into this one for some reason.
https://github.com/python-discord/bot/pull/2523/commits/e74e2e50469be60664afd564eab3a6d50961c0f4

File: .pre-commit-config.yaml
Comment (wookie184) [reply .pre-commit-config.yaml]: Looking at how `ruff-pre-commit` does it (https://github.com/charliermarsh/ruff-pre-commit/blob/f6346ef27faac608d493bf3926528b308d175d9e/.pre-commit-hooks.yaml) I believe precommit automatically passes the files to run on (only changed ones) so we don't need to specify to run on all files. Ruff is fast enough that it doesn't seem to make much of a difference anyway though :P

File: bot/constants.py
Comment (wookie184) [inline bot/constants.py:4]: Should be `superseded` not `superceded`, although I think `overridden` sounds better anyway.
```suggestion
By default, the values defined in the classes are used, these can be overridden by an env var with the same name.
```

**Review by wookie184 — CHANGES_REQUESTED** [review]

**Review by wookie184 — APPROVED** [review]
💯

**Review by jchristgit — APPROVED** [review]

**Comment by jchristgit** [comment]
BEAST!!!!!!

**Pull request merged**
